modules/fx/arrow-fx/src/main/kotlin/arrow/fx/IO.kt

```diff
@@ -1108,4 +995,148 @@ fun <A> IOOf<A>.handleError(f: (Throwable) -> A): IO<A> =
  * @see handleErrorWith for a version that can resolve the error using an effect
  */
 fun <A> IOOf<A>.handleErrorWith(f: (Throwable) -> IOOf<A>): IO<A> =
-  IO.Bind(fix(), IOFrame.Companion.ErrorHandler(f))
+  BIO.Bind(fix(), IOFrame.Companion.ErrorHandler(f))
+
+/**
+ * Executes the given `finalizer` when the source is finished, either in success or in error, or if canceled, allowing
+ * for differentiating between exit conditions. That's thanks to the [ExitCase] argument of the finalizer.
+ *
+ * As best practice, it's not a good idea to release resources via `guaranteeCase` in polymorphic code.
+ * Prefer [bracketCase] for the acquisition and release of resources.
+ *
+ * @see [guarantee] for the simpler version
+ * @see [bracketCase] for the more general operation
+ *
+ */
+fun <A> IOOf<A>.guaranteeCase(finalizer: (ExitCase<Throwable>) -> IOOf<Unit>): IO<A> =
+  IOBracket.guaranteeCase(fix(), finalizer)
+
+/**
+ * Transform the [IO] value of [A] by sequencing an effect [IO] that results in [B].
+ *
+ * @param f function that returns the [IO] effect resulting in [B] based on the input [A].
+ * @returns an effect that results in [B].
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.IO
+ *
+ * fun main(args: Array<String>) {
+ *   val result =
+ *   //sampleStart
+ *   IO.just("Hello").flatMap { IO { "$it World" } }
+ *   //sampleEnd
+ *   println(result.unsafeRunSync())
+ * }
+ * ```
+ */
+fun <E, A, B, E2 : E> BIOOf<E, A>.flatMap(f: (A) -> BIOOf<E2, B>): BIO<E2, B> =
+  when (val current = fix()) {
+    is BIO.RaiseException -> current
+    is BIO.RaiseError<E> -> current as BIO<E2, B>
+    is BIO.Pure<A> -> BIO.Suspend { f(current.a) }
+    else -> BIO.Bind(current, f)
+  }
+
+/**
+ * Meant for specifying tasks with safe resource acquisition and release in the face of errors and interruption.
+ * It would be the the equivalent of `try/catch/finally` statements in mainstream imperative languages for resource
+ * acquisition and release.
+ *
+ * @param release is the action that's supposed to release the allocated resource after `use` is done, irregardless
+ * of its exit condition.
+ *
+ * ```kotlin:ank:playground
+ * import arrow.fx.IO
+ *
+ * class File(url: String) {
+ *   fun open(): File = this
+ *   fun close(): Unit {}
+ *   override fun toString(): String = "This file contains some interesting content!"
+ * }
+ *
+ * fun openFile(uri: String): IO<File> = IO { File(uri).open() }
+ * fun closeFile(file: File): IO<Unit> = IO { file.close() }
+ * fun fileToString(file: File): IO<String> = IO { file.toString() }
+ *
+ * fun main(args: Array<String>) {
+ *   //sampleStart
+ *   val safeComputation = openFile("data.json").bracket({ file: File -> closeFile(file) }, { file -> fileToString(file) })
+ *   //sampleEnd
+ *   println(safeComputation.unsafeRunSync())
+ * }
+ * ```
+ */
+fun <A, B> IOOf<A>.bracket(release: (A) -> IOOf<Unit>, use: (A) -> IOOf<B>): IO<B> =
```

Comment:
Since both `IO` and `BIO` will implement all these methods, would it make sense to separate both in different files even if `IO` is an alias for `BIO`? We could keep here `IO` alias an all its combinators and move all BIO stuff to a separate file called `BIO.kt` so it's also easier to search. Feels like we'll have a big bunch of code within this file otherwise.

Reply:
Just like `flatMap` these will all be refactored to work for `BIO<E, A>` which means they'll work for both. This is what we'll be splitting over multiple PRs since it's a quite large diff. (~3k loc for refactoring all combinators and adding a couple new ones to work on `E`)

Reply:
Perfect. Much more clear now. It got me wondering why some methods were defined over `IO` and some others over `BIO`.